test/e2e/framework/helper/validations/certificates.go

```diff
@@ -0,0 +1,257 @@
+	}
+
+	return nil
+}
+
+// ExpectValidPrivateKeyData checks of the secret's private key matches the request
+func ExpectValidPrivateKeyData(certificate *cmapi.Certificate, secret *corev1.Secret) error {
+	keyBytes, ok := secret.Data[corev1.TLSPrivateKeyKey]
+	if !ok {
+		return fmt.Errorf("No private key data found for Certificate %q (secret %q)", certificate.Name, certificate.Spec.SecretName)
+	}
+	key, err := pki.DecodePrivateKeyBytes(keyBytes)
+	if err != nil {
+		return err
+	}
+
+	// validate private key is of the correct type (rsa or ecdsa)
+	if certificate.Spec.PrivateKey != nil {
+		switch certificate.Spec.PrivateKey.Algorithm {
+		case cmapi.PrivateKeyAlgorithm(""),
+			cmapi.RSAKeyAlgorithm:
+			_, ok := key.(*rsa.PrivateKey)
+			if !ok {
+				return fmt.Errorf("Expected private key of type RSA, but it was: %T", key)
+			}
+		case cmapi.ECDSAKeyAlgorithm:
+			_, ok := key.(*ecdsa.PrivateKey)
+			if !ok {
+				return fmt.Errorf("Expected private key of type ECDSA, but it was: %T", key)
+			}
+		default:
+			return fmt.Errorf("unrecognised requested private key algorithm %q", certificate.Spec.PrivateKey.Algorithm)
+		}
+	}
+
+	// TODO: validate private key KeySize
+	return nil
+}
+
+// ExpectValidCertificate checks if the certificate is a valid x509 certificate
+func ExpectValidCertificate(certificate *cmapi.Certificate, secret *corev1.Secret) error {
+	certBytes, ok := secret.Data[corev1.TLSCertKey]
+	if !ok {
+		return fmt.Errorf("No certificate data found for Certificate %q (secret %q)", certificate.Name, certificate.Spec.SecretName)
+	}
+
+	_, err := pki.DecodeX509CertificateBytes(certBytes)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// ExpectCertificateOrganizationToMatch checks if the issued certificate has the same Organization as the requested one
+func ExpectCertificateOrganizationToMatch(certificate *cmapi.Certificate, secret *corev1.Secret) error {
+	cert, err := pki.DecodeX509CertificateBytes(secret.Data[corev1.TLSCertKey])
+	if err != nil {
+		return err
+	}
+
+	expectedOrganization := pki.OrganizationForCertificate(certificate)
+	if !util.EqualUnsorted(cert.Subject.Organization, expectedOrganization) {
+		return fmt.Errorf("Expected certificate valid for O %v, but got a certificate valid for O %v", expectedOrganization, cert.Subject.Organization)
+	}
+
+	return nil
+}
+
+// ExpectCertificateDNSNamesToMatch checks if the issued certificate has all DNS names it requested
+func ExpectCertificateDNSNamesToMatch(certificate *cmapi.Certificate, secret *corev1.Secret) error {
+	cert, err := pki.DecodeX509CertificateBytes(secret.Data[corev1.TLSCertKey])
+	if err != nil {
+		return err
+	}
+
+	expectedDNSNames := certificate.Spec.DNSNames
+	if !util.Subset(cert.DNSNames, expectedDNSNames) {
+		return fmt.Errorf("Expected certificate valid for DNSNames %v, but got a certificate valid for DNSNames %v", expectedDNSNames, cert.DNSNames)
+	}
```

Comment:
Why use `util.Subset` here rather than `EqualUnsorted`? Under what circumstances would the the signed certificate have more DNS names than were requested? Add an explanatory note.

Reply:
Honest: i do not know... this is copied over... I guess if a CN got added to the DNS names